### ISSUE TITLE
compatible with "snap" firefox?

### ISSUE THREAD

**Issue by jhnc** [jlebon/textern#81]
When updating from Ubuntu 20.04 to 22.04, firefox changes from traditional package to a "snap" package.

textern's `make native-install USER=1` option assumes that installation should be to `~/.local` and `~/.mozilla/...`. However, with "snap", the paths are different (and I guess much more).

I see many reports of other software breaking that needs to interact with "snap" or "flatpak" programs.

Is it expected that textern works with "snap" firefox? If so, what is the installation procedure?

**Comment by jlebon** [comment]
I don't have a system to test this, but if you or someone finds out the right path to use, I'd be happy to accept a patch with possible tweaks to the Makefile and appropriate docs.

**Comment by FelixJacobi** [comment]
Not, it is not, as Native Messaging does not work due the Snap sandbox (Firefox is not allowed to start any executable outside of its sandbox), so it cannot access the editor executable on your system. It would work, if you would install the editor and Textern into the Snap image, which is not possible in practice.

There is in issue in Bugzilla tracking the missing support: [1661935 - Snap does not support NativeMessaging](https://bugzilla.mozilla.org/show_bug.cgi?id=1661935)

**Comment by FelixJacobi** [comment]
Update: With the recent update of the Firefox snap, it should work now 🎉  .

To make it work:

* The Firefox snap does not use `~/.mozilla` anymore, the settings directory is now located in `~/snap/firefox/common/.mozilla`.
* I would recommend creating a symlink on the old location:

~~~
mkdir -p ~/.mozilla
cd ~/,mozilla
ln -s ../snap/firefox/common/.mozilla/firefox
ln -s ../snap/firefox/common/.mozilla/native-messaging-hosts
~~~

Run `make native-install USER=1` again, restart Firefox and the next time when you press the keyboard shortcut to launch the editor, you should get a permission prompt from your desktop environment to allow Firefox to run the external executable for native messaging.

**Comment by jhnc** [comment]
Thanks! I got it to work with 107.0-2/2088 using @FelixJacobi 's instructions.

I use `xfce4-terminal` which seems to need the `--disable-server` option.

Is it expected that empty directories accumulate in `/run/user/$(id -u)/textern/` ?

**Comment by FelixJacobi** [comment]
> I had to manually create `~/snap/firefox/common/.mozilla/native-messaging-hosts/`.

Thanks, missed that. I've updated my comment to include these.

**Comment by jlebon** [comment]
Instead of creating the symlinks, does it work if you do

```
$ make native-install USER=1 MOZILLA_NATIVE=$HOME/.mozilla/native-messaging-hosts
```

?

**Comment by FelixJacobi** [comment]
> Instead of creating the symlinks, does it work if you do
> 
> ```
> $ make native-install USER=1 MOZILLA_NATIVE=$HOME/.mozilla/native-messaging-hosts
> ```
> 
> ?

Yeah, could work also. I created the symlink, as there are some other applications utilizing native messaging and assuming that `native-messaging-hosts` is located in `~/.mozilla`.

**Comment by jlebon** [comment]
Ack makes sense. Would appreciate a patch to add instructions for Firefox snaps if anyone is interested (maybe at the same level as https://github.com/jlebon/textern#firejail). Extra points to also cover snaps of editors if it requires any special knowledge (maybe at the same level as https://github.com/jlebon/textern#flatpak). :)

**Comment by dmusican** [comment]
Like @chubin, I've followed the above instructions, and I'm unable to get textern to work with snap Firefox. Here is the error I see in the console:

```
NS_ERROR_ILLEGAL_VALUE: Component returned failure code: 0x80070057 (NS_ERROR_ILLEGAL_VALUE) [nsINativeMessagingPortal.closeSession]
```

**Comment by jbuttery** [comment]
I've also made an attempt to get Textern working with a Firefox snap; success has so far eluded me, but my symptoms are different.  I adjusted the Makefile:

```Makefile
PREFIX ?= $(HOME)/snap/firefox/current/.local
MOZILLA_NATIVE ?= $(HOME)/snap/firefox/current/.mozilla/native-messaging-hosts
```

...then ran the installation:

```sh
make native-install USER=1
```

...but that hasn't done the trick.  When I press `Ctrl+Shift+d` there's no response on the page; I do have this in the extension's debug log:

```
Unable to check <input pattern=‘[.a-zA-Z0-9_-]+’> because ‘/[.a-zA-Z0-9_-]+/v’ is not a valid regexp: invalid character in class in regular expression
```

The above line appears once each for:

- `[options.js:40:18](moz-extension://265f90e3-d0e7-4690-9b02-30a8358a30ac/options.js)`
- `[options.html](moz-extension://265f90e3-d0e7-4690-9b02-30a8358a30ac/options.html)`

The regex _seems_ fine (I checked it with https://regex101.com/ against the default value of `txt`), but I don't recognize the `v` flag at the end.  I did change the `-` at the end to `\-` to really drive home the point that it's not a range operator, but that shouldn't be necessary.

Anyway, those errors seem to appear when the preferences page loads, so I'm not sure they're really relevant here.

What's definitely relevant is that whenever I hit `Ctrl+Shift+d` in a textarea, this shows up in the extension debug log:

```
Disconnected from helper [background.js:83:21](moz-extension://265f90e3-d0e7-4690-9b02-30a8358a30ac/background.js)
Error: Error: An unexpected error occurred [background.js:14:13](moz-extension://265f90e3-d0e7-4690-9b02-30a8358a30ac/background.js)
```

Does Textern write its own logs anywhere, that might have more detail on what's going on?

Here's another interesting fact: after I'd poked at this for a while and couldn't get it to work, I thought maybe I'd need to restart Firefox.  I did, and when the browser windows closed, they revealed a dialog about allowing Textern native messaging.  UGH...UI fail!  I clicked OK but I don't know if the Firefox mothership was still listening or not.  I haven't been able to get that dialog to show up again (and I'm periodically minimizing Firefox' windows to look for it now), even after deleting the `native-messaging-hosts` directories (in my home directory and in the snap directory).  I'm gonna be super disappointed if that winds up being what tripped all this up.

Anyway,  I'm gonna go poke some more...but if this suggests any further debugging routes, feel free to let me know.

**Comment by jhnc** [comment]
> The regex _seems_ fine (I checked it with https://regex101.com/ against the default value of `txt`), but I don't recognize the `v` flag at the end. I did change the `-` at the end to `\-` to really drive home the point that it's not a range operator, but that shouldn't be necessary.

See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Character_class#v-mode_character_class

The backslash-escaping the hyphen is necessary. If you've added it, and the regex is still giving errors, something else must be wrong with it.
